### PR TITLE
Update srcset / sizes to match browser behavior experiment

### PR DESCRIPTION
**Pull request by cathyxz**
Implements an experiment to test addressing https://github.com/ampproject/amphtml/issues/11575. I've addressed all the edge cases I can think of, but there are probably more that I'm not yet considering. 
1. We need to propagate a `src` for browsers that don't support `srcset` (IE). 
2. We need to avoid breaking the current `sizes` behavior (achieve this by using `native-sizes`). 

- [x] Propagate srcset and sizes from `<amp-img>` to child ``. 
- [x] Add `native-sizes` attribute to `<amp-img>`. 
- [x] Update Validator wrt `native-sizes` and `sizes`. 
- [x] Verify via manual tests
- [x] Add unit tests
- [x] Review documentation (verify is correct)
- [x] Gate this behind an experiment
- [x] Review potential points of breakage / edge cases
- [x] Test IE for `src` undefined edge case. 
- [x] Address comments

### CODE REVIEW

File: builtins/amp-img.js [hunk 1]
```diff
@@ -141,6 +145,17 @@ export class AmpImg extends BaseElement {
     }
 
     this.propagateAttributes(ATTRIBUTES_TO_PROPAGATE, this.img_);
+
+    if (this.useNativeSrcset_) {
+      this.propagateAttributes(['src', 'srcset', 'sizes'], this.img_);
+
+      const nativeSizes = this.element.getAttribute('native-sizes');
```
Comment (cramforce) [inline builtins/amp-img.js]: On link-rel-preload this is called `imgsizes` because `sizes` means something else there. Lets use the same name.

File: builtins/amp-img.js [hunk 2]
```diff
@@ -141,6 +145,17 @@ export class AmpImg extends BaseElement {
...
+      if (nativeSizes) {
+        // Otherwise sizes will be propagated normally if defined
+        this.img_.setAttribute('sizes', nativeSizes);
```
Comment (cramforce) [inline builtins/amp-img.js]: I'm slightly puzzled by this code, but there might be a misunderstanding on the goal. When 'amp-img-native-srcset' is set we should never even initialize `this.srcset_`. And we talked about generating a sizes attribute if `native-sizes` is not provided, so the browser can choose the right one from the `srcset` based on screen size only (emulating the behavior of our current implementation).

File: builtins/amp-img.js [hunk 3]
```diff
@@ -197,6 +213,13 @@ export class AmpImg extends BaseElement {
 
     this.img_.setAttribute('src', src);
 
+    if (this.useNativeSrcset_) {
```
Comment (cramforce) [inline builtins/amp-img.js]: Don't do all the work above if going into this branch. I expected that you needed to generated a `sizes` attribute if the user provided none.

File: test-amp-img.js [hunk 4]
```diff
@@ -0,0 +1,486 @@
+/**
```
Comment (cathyxz) [inline test-amp-img.js]: TODO: remove this file. Git did something weird here...

File: builtins/amp-img.js (hunk 2 above)
Comment (cathyxz) [reply builtins/amp-img.js]: I preserved this code because of the edge case where `srcset` is not supported (IE, Opera Mini) but the user provided `srcset` instead of `src`. But I think you are right, we should not initialize `srcset` at all, and it probably suffices to just persist the first url in `srcset` like we currently do in the preconnect callback. 

Automatically generating `sizes` is an outstanding FR as part of this issue, just not addressed in this PR.

Comment (cramforce) [reply builtins/amp-img.js]: I'm not sure supporting browsers without srcset support is worth it anymore since the fallback to just src is fine.

If you wanted to do that, then we should only call the code if `'srcset' in img` is false, but not do the work and then override it.

File: builtins/amp-img.js (hunk 3 above)
Comment (aghassemi) [reply builtins/amp-img.js]: +1 even better if we only parse and create `srcset_` if 1- `src` is not provided and 2- browser does not support `srcset` (even if the check is a simple if IE <= 11)

File: validator/validator-main.protoascii [hunk 5]
```diff
@@ -1887,6 +1887,7 @@ tags: {
   attrs: { name: "border" }
   attrs: { name: "height" }
   attrs: { name: "width" }
+  attrs: { name: "sizes" }
```
Comment (aghassemi) [inline validator/validator-main.protoascii]: `sizes` is already a universal attribute in AMP and given we don't need `"native-sizes"` per comment above, there shouldn't be a need for validation changes.

File: builtins/amp-img.js [hunk 6]
```diff
@@ -26,6 +27,15 @@ import {srcsetFromElement, srcsetFromSrc} from '../src/srcset';
...
   'aria-describedby', 'aria-labelledby'];
 
+
+const EXPERIMENTAL_ATTRIBUTES = ['srcset', 'src', 'sizes'];
+
+/**
+ * Regex for finding the first url in a srcset.
+ * @type {!RegExp}
+ */
+const FIRST_SRCSET_URL = /https?:\/\/\S+/;
```
Comment (aghassemi) [inline builtins/amp-img.js]: This won't handle relative urls which can happen on origin (cache rewrites all images). Probably safer to import `srcsetRegex` from `srcset.js` and grab the first item. (we still end up reducing the v0.js a lot anyway since most of code from `srcset.js` won't get imported after this is fully launched.

Comment (aghassemi) [reply builtins/amp-img.js]: (this regex was fine for preconnect case since that only matters for cache cases where img Urls are fully rewritten anyway)

File: builtins/amp-img.js [hunk 7]
```diff
@@ -178,6 +203,21 @@ export class AmpImg extends BaseElement {
...
+   * Sets the img src to the first url in the srcset if srcset is defined but
+   * src is not.
+   * @private
+   */
+  guaranteeSrcForSrcsetUnsupportedBrowsers_() {
+    if (!this.img_.hasAttribute('src') && 'srcset' in this.img_ == false) {
+      const srcset = this.element.getAttribute('srcset');
+      const srcseturl = srcsetRegex.exec(srcset);
+      if (srcseturl) {
+        this.img_.setAttribute('src', srcseturl[0]);
```
Comment (aghassemi) [inline builtins/amp-img.js]: `match` would return an array of arrays. So [0] would be an array itself. First item should the the whole line match (e.g. `file.png 200w`) [1] should be the first grouping which is the URL and what we want. let's add a test for it as well.

File: builtins/amp-img.js [hunk 8]
```diff
@@ -26,6 +27,10 @@ import {srcsetFromElement, srcsetFromSrc} from '../src/srcset';
 const ATTRIBUTES_TO_PROPAGATE = ['alt', 'title', 'referrerpolicy', 'aria-label',
   'aria-describedby', 'aria-labelledby'];
 
+const EXPERIMENTAL_ATTRIBUTES_TO_PROPAGATE = ['alt', 'title',
```
Comment (jridgewell) [inline builtins/amp-img.js]: Use `concat`?

File: builtins/amp-img.js (hunk 7 above)
Comment (jridgewell) [reply builtins/amp-img.js]: It uses `RegExp#exec`, not `match`. 😉

But, you do want the url at `[1]`, not the entire source at `[0]`.

File: builtins/amp-img.js [hunk 9]
```diff
@@ -178,6 +203,21 @@ export class AmpImg extends BaseElement {
...
 
+  /**
+   * Sets the img src to the first url in the srcset if srcset is defined but
+   * src is not.
+   * @private
+   */
+  guaranteeSrcForSrcsetUnsupportedBrowsers_() {
+    if (!this.img_.hasAttribute('src') && 'srcset' in this.img_ == false) {
+      const srcset = this.element.getAttribute('srcset');
+      const srcseturl = srcsetRegex.exec(srcset);
```
Comment (jridgewell) [inline builtins/amp-img.js]: OOOOoooo, @aghassemi's comment reminded me of something else:

Non-exhaustive use of global regexes is dangerous! Global regexes have state (`lastIndex`) which is used to find the next match in a series. `#match` is exhaustive, so it's fine. But, `#exec` isn't, so it leaves a new `lastIndex` on the regex after matching, causing mayhem!

```js
srcsetRegex.lastIndex; // => 0
srcsetRegex.exec('test.jpg'); // => ["test.jpg", "test.jpg", undefined, undefined]

srcsetRegex.lastIndex; // => 8
srcsetRegex.exec('test2.jpg'); // => ["g", "g", undefined, undefined]
```

You have to manually reset the `lastIndex` yourself afterwards! Or better yet, export a helper method that'll do the `#exec` and reset it for you.

Comment (aghassemi) [reply builtins/amp-img.js]: wow, TIL & scary! (current ToT about making interfaces predictable perfectly applies here)

Comment (cathyxz) [reply builtins/amp-img.js]: oh wow.

File: builtins/amp-img.js (hunk 7 above)
Comment (cathyxz) [reply builtins/amp-img.js]: Good catch!

Comment (cathyxz) [reply builtins/amp-img.js]: So if we think about this, we really only need the first url in the `srcset`. It's a url that can be `http`, `https` or relative, and ends with a space. So can we just match the first `\S+` and be done with it? If we wanted to validate more, we could do `(http|https|\/)\S+`.

File: builtins/amp-img.js [hunk 10]
```diff
@@ -178,6 +202,22 @@ export class AmpImg extends BaseElement {
...
+  /**
+   * Sets the img src to the first url in the srcset if srcset is defined but
+   * src is not.
+   * @private
+   */
+  guaranteeSrcForSrcsetUnsupportedBrowsers_() {
+    if (!this.img_.hasAttribute('src') && 'srcset' in this.img_ == false) {
+      const srcset = this.element.getAttribute('srcset');
+
+      const srcseturl = /\S+/.match(srcset);
```
Comment (aghassemi) [inline builtins/amp-img.js]: nice simplification. LGTM, can't think of cases this may fail, @jridgewell WDYT?

Comment (jridgewell) [reply builtins/amp-img.js]: Yah, this'll work.

**Comment by cathyxz** [comment]
Looks like I need to increase size cap here. =(
>FAIL  ./dist/amp.js: 337.39KB > maxSize 337.07KB (gzip) 

But the increase here is all due to experiment code. Either way, once the experiment is over, the net code change will be a reduction.

**Comment by jridgewell** [comment]
Rebase on master.

**Comment by cathyxz** [comment]
This is after merging latest master: 
>FAIL  ./dist/v0.js: 77.21KB > maxSize 77.15KB (gzip)

**Comment by jridgewell** [comment]
> FAIL ./dist/v0.js: 77.21KB > maxSize 77.15KB (gzip)

All good, bump up the limit.

File: build-system/tasks/bundle-size.js [hunk 11]
```diff
@@ -21,7 +21,7 @@ const gulp = require('gulp-help')(require('gulp'));
 const log = require('fancy-log');
 
 const runtimeFile = './dist/v0.js';
-const maxSize = '77.15KB';
+const maxSize = '77.21KB';
```
Comment (rsimha) [inline build-system/tasks/bundle-size.js]: The very next build on `master` went past this limit :(

https://travis-ci.org/ampproject/amphtml/jobs/375136775#L641

Looks like we need to provide a larger buffer :/
